### PR TITLE
Feature/upsert endpoint

### PR DESCRIPTION
**Pull request by southeo**
Introduces upsert endpoint that can create and update handles. Endpoint is PATCH `/api/v1/pids/upsert`

Checks if a handle for a given digital specimen already exists by checking primarySpecimenObjectId. Splits request into two components: new and updates. For the specimens to update, the issueNumber is incremented. 

This also includes code from the approved PRs "better archiving" and "rollback endpoints."

### CODE REVIEW

File: src/test/java/eu/dissco/core/handlemanager/controller/HandleControllerTest.java [hunk 1]
```diff
@@ -122,6 +108,12 @@ void testResolveSingleHandle() throws Exception {
     assertThat(responseReceived.getBody()).isEqualTo(responseExpected);
   }
 
+  @Test
+  void testResolvePidBadPrefix(){
+    assertThrows(PidResolutionException.class, () ->
+        controller.resolvePid(SUFFIX, SUFFIX, new MockHttpServletRequest()));
```
Comment (samleeflang) [inline src/test/java/eu/dissco/core/handlemanager/controller/HandleControllerTest.java:114]: Shouldn't this be `PREFIX, SUFFIX` ?

**Review by samleeflang — COMMENTED** [review]
Shaping up great, looking forward to see it in action.
I think we could use some more log messages, it will definitely help us when something goes wrong in the future.
I was also wondering if you have any idea how performant it is?

File: src/main/java/eu/dissco/core/handlemanager/service/ServiceUtils.java [hunk 2]
```diff
@@ -3,32 +3,45 @@
...
 
-  private static byte[] concatIds(String physicalIdentifier, String specimenHostPid) {
+  private static String concatIds(String physicalIdentifier, String specimenHostPid) {
     var hostIdArr = specimenHostPid.split("/");
     var hostId = hostIdArr[hostIdArr.length - 1];
-    return (physicalIdentifier + ":" + hostId).getBytes(StandardCharsets.UTF_8);
+    return (physicalIdentifier + ":" + hostId);
+  }
+
+  public static <T> Collector<T, ?, T> toSingleton() {
```
[the remark this answers is not in the document]
Comment (southeo) [reply src/main/java/eu/dissco/core/handlemanager/service/ServiceUtils.java:35]: is toSingleObject okay, or is it too close?

File: src/test/java/eu/dissco/core/handlemanager/controller/HandleControllerTest.java (hunk 1 above)
Comment (southeo) [reply src/test/java/eu/dissco/core/handlemanager/controller/HandleControllerTest.java:113]: It's testing how the program reacts when the prefix is bad (i.e. not equal to 20.500.1025).

File: src/main/java/eu/dissco/core/handlemanager/controller/HandleController.java [hunk 3]
```diff
@@ -172,7 +153,22 @@ public ResponseEntity<JsonApiWrapperWrite> updateRecords(@RequestBody List<JsonN
...
-    return ResponseEntity.status(HttpStatus.OK).body(service.updateRecords(requests));
+    return ResponseEntity.status(HttpStatus.OK).body(service.updateRecords(requests, true));
+  }
+
+  // Upsert
+  @Operation(summary="Create a PID Record; if it already exists, update contents. DigitalSpecimens only.")
+  @PatchMapping(value="/upsert")
+  public ResponseEntity<JsonApiWrapperWrite> upsertRecord(@RequestBody List<JsonNode> requests)
+      throws InvalidRequestException, UnprocessableEntityException, PidResolutionException, PidServiceInternalError, JsonProcessingException {
+    log.info("received upsert request");
```
[the remark this answers is not in the document]
Comment (southeo) [reply src/main/java/eu/dissco/core/handlemanager/controller/HandleController.java]: Added logging info

File: src/main/java/eu/dissco/core/handlemanager/controller/HandleController.java [hunk 4]
```diff
@@ -194,6 +190,32 @@ public ResponseEntity<JsonApiWrapperWrite> archiveRecord(@PathVariable("prefix")
...
+  @DeleteMapping(value="/rollback")
+  public ResponseEntity<Void> rollbackHandleCreation(@RequestBody RollbackRequest request, Authentication authentication)
+      throws InvalidRequestException {
+    var ids = request.data().stream().map(d -> d.get(NODE_ID)).toList();
+    if (ids.contains(null)){
+      throw new InvalidRequestException("Missing Handles (\"id\") in request");
+    }
+    var handles = ids.stream().map(JsonNode::asText).toList();
+
+    log.info("Rollback request received from user "+ authentication.getName() + " for handles : " + handles);
```
[the remark this answers is not in the document]
Comment (southeo) [reply src/main/java/eu/dissco/core/handlemanager/controller/HandleController.java]: Fixed?

File: src/main/java/eu/dissco/core/handlemanager/service/ServiceUtils.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (southeo) [reply src/main/java/eu/dissco/core/handlemanager/service/ServiceUtils.java]: Removed this altogether

File: src/test/java/eu/dissco/core/handlemanager/controller/HandleControllerTest.java (hunk 1 above)
Comment (samleeflang) [reply src/test/java/eu/dissco/core/handlemanager/controller/HandleControllerTest.java:122]: :+1:

File: pom.xml [hunk 5]
```diff
@@ -86,6 +86,20 @@
       <groupId>org.springframework.boot</groupId>
       <artifactId>spring-boot-starter-validation</artifactId>
     </dependency>
+    <dependency>
+      <groupId>org.springframework.boot</groupId>
+      <artifactId>spring-boot-configuration-processor</artifactId>
+      <optional>true</optional>
+    </dependency>
+    <dependency>
```
Comment (samleeflang) [inline pom.xml:94]: To me it looks like the configuration-processor is in there twice

File: src/main/java/eu/dissco/core/handlemanager/component/FdoRecordBuilder.java [hunk 6]
```diff
@@ -241,35 +267,144 @@ public List<HandleAttribute> prepareDoiRecordAttributes(DoiRecordRequest request
...
+    var fdoRecord = prepareDoiRecordAttributes(request, handle, type);
 
-    // 14 Media Hash
+    // 400 MediaHash
     fdoRecord.add(new HandleAttribute(FIELD_IDX.get(MEDIA_HASH), handle, MEDIA_HASH,
         request.getMediaHash().getBytes(StandardCharsets.UTF_8)));
 
-    // 15 Subject Specimen Host
+    // 401 mediaHashAlgorithm
+    fdoRecord.add(
```
Comment (samleeflang) [inline src/main/java/eu/dissco/core/handlemanager/component/FdoRecordBuilder.java]: Oh damn, I don't think we are going to hash media... It shouldn't be in here, sorry will pick this up with Wouter. Was there an RFC for images?

Comment (samleeflang) [reply src/main/java/eu/dissco/core/handlemanager/component/FdoRecordBuilder.java]: Leave it in for now

File: src/main/java/eu/dissco/core/handlemanager/component/FdoRecordBuilder.java [hunk 7]
```diff
@@ -445,28 +580,61 @@ private List<HandleAttribute> setSpecimenHostName(DigitalSpecimenRequest request
...
+  private List<HandleAttribute> addResolvedNames(Map<String, String> updateRequestMap, byte[] handle)
+      throws UnprocessableEntityException, PidResolutionException {
+
+    var resolvableKeys = updateRequestMap.entrySet()
+        .stream()
+        .filter(entry -> RESOLVABLE_KEYS.containsKey(entry.getKey())
+            && !hasResolvedPairInRequest(updateRequestMap, entry.getKey()))
+        .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+    ArrayList<HandleAttribute> resolvedPidNameAttributes = new ArrayList<>();
```
Comment (samleeflang) [inline src/main/java/eu/dissco/core/handlemanager/component/FdoRecordBuilder.java:615]: Would put this on top of the for loop where it is used

File: src/main/java/eu/dissco/core/handlemanager/controller/HandleController.java [hunk 8]
```diff
@@ -52,177 +54,211 @@
...
-  @Operation(summary ="Create multiple PID Records at a time.")
+  @Operation(summary = "Create multiple PID Records at a time.")
   @PostMapping(value = "/batch")
-  public ResponseEntity<JsonApiWrapperWrite> createRecords(@RequestBody List<JsonNode> requests)
+  public ResponseEntity<JsonApiWrapperWrite> createRecords(@RequestBody List<JsonNode> requests,
+      Authentication authentication)
       throws PidResolutionException, PidServiceInternalError, InvalidRequestException, PidCreationException {
-    log.info("received batch request");
-
+    log.info("Validating batch POST request from user {}", authentication);
```
Comment (samleeflang) [inline src/main/java/eu/dissco/core/handlemanager/controller/HandleController.java:123]: Probably better to get the name here

File: src/main/java/eu/dissco/core/handlemanager/controller/HandleController.java [hunk 9]
```diff
@@ -52,177 +54,211 @@
...
     schemaValidator.validatePutRequest(request);
 
     JsonNode data = request.get(NODE_DATA);
     byte[] handle = (prefix + "/" + suffix).getBytes(StandardCharsets.UTF_8);
     byte[] handleRequest = data.get(NODE_ID).asText().getBytes(StandardCharsets.UTF_8);
     if (!Arrays.equals(handle, handleRequest)) {
       throw new InvalidRequestException(
-          "Handle in request URL does not match id in request body. URL: " + handle + ", body: "
-              + handleRequest);
+          String.format(
```
Comment (samleeflang) [inline src/main/java/eu/dissco/core/handlemanager/controller/HandleController.java:209]: Might want to stick with the concatenation. String.format is very slow. However, here it is not really an issue

**Review by samleeflang — COMMENTED** [review]
I'm impressed this is looking very good. Couple small comments

File: src/main/java/eu/dissco/core/handlemanager/controller/HandleController.java (hunk 8 above)
Comment (southeo) [reply src/main/java/eu/dissco/core/handlemanager/controller/HandleController.java:123]: Good catch

File: src/main/java/eu/dissco/core/handlemanager/controller/HandleController.java (hunk 9 above)
Comment (southeo) [reply src/main/java/eu/dissco/core/handlemanager/controller/HandleController.java]: Reverted

**Review by samleeflang — APPROVED** [review]
:tada:

**Pull request merged**